### ISSUE TITLE
OpenSSL 1.1.0g error when client certificate uses EC public key.

### ISSUE THREAD

**Issue by clinew** [inspircd/inspircd#1464]
**Description**
The server is set up to only accept clients with trusted certificates.  In this case, the client has a certificate whose public key is an elliptic curve (specifically, `brainpoolP512t1`).  In this case the protocol will error.

**Steps to reproduce the issue:**
Generate a client certificate with an EC public key and try to connect with it.
```
openssl ecparam -out ec.pem -name brainpoolP512t1 -genkey
openssl req -x509 -new -key ec.pem -out cert.pem -subj "/CN=EC-SHA512/"
openssl s_client -cert cert.pem -key ec.pem -connect 127.0.0.1:6697
```

**Describe the results you received:**
Client disconnects from a protocol error.
```
140379411924736:error:1409441A:SSL routines:ssl3_read_bytes:tlsv1 alert decode error:ssl/record/rec_layer_s3.c:1399:SSL alert number 50
```

**Describe the results you expected:**
Client connects successfully.

**Additional information you deem important (e.g. issue happens only occasionally):**
I'm going to edit this as I find out more.

- The error did *not* occur on OpenSSL 1.0.1n.
- OpenSSL 1.1.0g is compiled with `--disable-deprecated`.
- (edit 01) Tested with OpenSSL 1.1.0g with `--enable-deprecated` and had same issue.
- Setting log level to `debug` and all input gives:
```
Sat Mar 03 2018 02:00:15 SOCKET: Accepting connection on socket 0.0.0.0:6697 fd 8
Sat Mar 03 2018 02:00:15 CULLLIST: classbase::+ @0x55ebc017ca30
Sat Mar 03 2018 02:00:15 USERS: New UUID for user: 474AAAAAA
Sat Mar 03 2018 02:00:15 CULLLIST: classbase::+ @0x55ebc017cc58
Sat Mar 03 2018 02:00:15 USERS: New user fd: 8
Sat Mar 03 2018 02:00:15 SOCKET: New file descriptor: 8
Sat Mar 03 2018 02:00:15 CULLLIST: classbase::+ @0x55ebc01563b0
Sat Mar 03 2018 02:00:15 CONNECTCLASS: Setting connect class for UID 474AAAAAA
Sat Mar 03 2018 02:00:15 CONNECTCLASS: Checking prereg
Sat Mar 03 2018 02:00:16 SOCKET: Error on FD 8 - 'Read Error'
Sat Mar 03 2018 02:00:16 USERS: QuitUser: 474AAAAAA=474AAAAAA 'Read Error'
Sat Mar 03 2018 02:00:16 SOCKET: DoWrite on errored or closed socket
Sat Mar 03 2018 02:00:16 CULLLIST: classbase::~ @0x55ebc01563b0
Sat Mar 03 2018 02:00:16 SOCKET: Remove file descriptor: 8
Sat Mar 03 2018 02:00:16 CULLLIST: Deleting @0x55ebc017ca30
Sat Mar 03 2018 02:00:16 CULLLIST: classbase::- @0x55ebc017cc58
Sat Mar 03 2018 02:00:16 CULLLIST: classbase::- @0x55ebc017ca30
Sat Mar 03 2018 02:00:16 CULLLIST: classbase::~ @0x55ebc017cc58
Sat Mar 03 2018 02:00:16 CULLLIST: classbase::~ @0x55ebc017ca30
```

**Output of `./bin/inspircd --version`:**
```
test ~/temp # /usr/bin/inspircd --version

InspIRCd-3.0.0-a6
```
27e92622b5655c3ecf3182bc511adfb8a07b8722

**Comment by SaberUK** [comment]
I can confirm the existence of this. It also affects the 2.0 branch.

**Comment by SaberUK** [comment]
I've added some more debugging calls to m_ssl_openssl and this seems to be the root cause:

>140735672292160:error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher:ssl/statem/statem_srvr.c:1404:

**Comment by clinew** [comment]
I don't recall seeing the issue in 2.0.25.  It cropped up when running https://github.com/clinew/inspircdtests/blob/master/ssl_strongcert.sh on 3.0 only because the test didn't understand the result.  I'll have to take another look to confirm sometime.

**Comment by clinew** [comment]
This is definitely not happening in 2.0.26 for me, only on 3.0 and only with the 1.1.0 series.

**Comment by SaberUK** [comment]
Having looked into this it seems to only be an issue with brainpoolP512t1 not any other curve like prime256v1. I will investigate further.

**Comment by SaberUK** [comment]
It seems like someone else ran into the same issue as you with something else.

https://github.com/openssl/openssl/issues/6332#issuecomment-391279142

As they mentioned this is not a bug so i'll be closing this now. Sorry it took me so long to look into it.

**Comment by clinew** [comment]
So things are a bit weirder here than expected, but the behaviour seems to mirror what I'm seeing in InspIRCd:

### Set-up
```
openssl ecparam -out eckey.pem -name brainpoolP512t1 -genkey
openssl req -x509 -new -key eckey.pem -out eccert.pem -subj "/CN=EC-SHA512/"
openssl genrsa -out rsakey.pem 4096
openssl req -key rsakey.pem -new -x509 -days 7200 -sha512 -out rsacert.pem -subj '/CN=RSA/'
```

### OpenSSL 1.1.0h
BREAKS (expected)
```
openssl s_server -accept 415 -cert eccert.pem -key eckey.pem -4
openssl s_client -connect 127.0.0.1:415
```

WORKS (expected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

WORKS (expected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4 -CAfile eccert.pem
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

BREAKS (unexpected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4 -CAfile eccert.pem -verify 9
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

### OpenSSL 1.0.2o
BREAKS (expected)
```
openssl s_server -accept 415 -cert eccert.pem -key eckey.pem -4
openssl s_client -connect 127.0.0.1:415
```

WORKS (expected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

WORKS (expected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4 -CAfile eccert.pem
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

WORKS (expected)
```
openssl s_server -accept 415 -cert rsacert.pem -key rsakey.pem -4 -CAfile eccert.pem -verify 9
openssl s_client -cert eccert.pem -key eckey.pem -connect 127.0.0.1:415
```

TL;DR: The server cert always breaks but the client only breaks on 1.1.0 when server peer verification is set.
